File: sanghoon/programmers-2/쿼드압축 후 개수세기.py
def solution(arr):
    answer = [0, 0]
    length = len(arr)

    def compression(a, b, l):
        start = arr[a][b]
        for i in range(a, a + l):
            for j in range(b, b + l):
                if arr[i][j] != start:
                    l = l // 2
                    compression(a, b, l)
                    compression(a, b + l, l)
                    compression(a + l, b, l)
                    compression(a + l, b + l, l)
                    return

        answer[start] += 1

    compression(0, 0, length)

    return answer

arr = [[1, 1, 0, 0], [1, 0, 0, 0], [1, 0, 0, 1], [1, 1, 1, 1]]
solution(arr)

File: sanghyun/programmers-2/쿼드압축 후 개수 세기.java
import java.util.Arrays;

class Solution {
    int[] answer = new int[2];
    public int[] solution(int[][] arr) {

        cycle(arr,0,arr.length,0,arr[0].length);

        return answer;
    }
    public void cycle(int[][] arr, int sX, int eX, int sY, int eY){
        boolean cheak = true;
        for1 :
        for (int i = sX; i < eX; i++) {
            for (int j = sY; j < eY; j++) {
                if(arr[sX][sY] != arr[i][j]){
                    cheak = false;
                    break for1;
                }
            }
        }
        if(cheak){
            answer[arr[sX][sY]]++;
            return;
        }
        cycle(arr,sX,(sX+eX)/2,sY,(sY+eY)/2);//1
        cycle(arr,sX,(sX+eX)/2,(sY+eY)/2,eY);//3
        cycle(arr,(sX+eX)/2,eX,sY,(sY+eY)/2);//2
        cycle(arr,(sX+eX)/2,eX,(sY+eY)/2,eY);//4
    }
}
